Add mps command for managing message pickup stations

jboverfelt opened the pull request:
ref #51

This implements mps updating (add, delete, list, and list all) using the Winlink API. It takes explicit style inspiration from the account management functions already implemented. It doesn't appear that Winlink actually enforces this, but the API docs say there is a 30 minute rate limit on listing all MPSes so I honored that here. I managed to update and list my own MPS stations successfully over the last few days when testing this.

I did not attempt to implement using an INQUIRY message as described in #51 as I wasn't sure how you would want that to be handled. Perhaps we might use an "--as-message" flag in addition to one of the other options to automatically post to the user's outbox. I do think it would be good to implement as it would allow MPS management without an internet connection.

I also did not attempt to integrate this into the web ui as I wanted to make sure you were happy with the command line implementation first.

Justin AB3E

martinhpedersen commented:
Hi Justin,

Thanks for submitting this PR 😄 I appreciate your decision to focus on the CLI initially, and only via their web API.

I haven't done a full review yet, but it looks very promising!

I'd like to discuss the CLI command/option design. In the newly added `templates` and `account` commands I have introduced subcommand using the `func shiftArgs`. I'm wondering if that would be a better UX for this, in combination with a few command options. I believe it also simplifies the implementation, as we no longer need to count and check for number of options == 1.

```
pat mps list [--all] (defaults to listing mine)
pat mps add W1AW
pat mps delete
```

What do you think? Would it be more intuitive? 🤷 

I'm also wondering if we should rename `delete` to either `clear` or `delete-all` to better communicate that this clears the entire list, and can't be used to delete a single entry. We should probably also print the current list before deleting it, so the user can rebuild their list manually if they accidentally deletes it.

As for the rate limit, I think we need to cache the list so the user can retrieve a copy of the list again without waiting the 30 minutes. But let's focus on the command interface first 😊

jboverfelt commented:
Didn't mean to push - still working

jboverfelt commented:
Ok it's ready for review. I realized something in making these changes - "list all" is really a list of everyone's message pickup station configurations, not a list of available MPS. So I changed the behavior to output just the MPS callsigns and a count of how many people use it just like Winlink Express does. 

Changes:

- Rebased on develop, and updated CONTRIBUTING.md to make it clear that's what is desired
- Updated command options
- Changed the list --all behavior to output just MPS callsigns and user counts
- Changed delete to clear to make it obvious that it clears all stations
- Print out previous stations after a clear operation

I didn't implement the caching as discussed. I also felt the test wasn't really providing a ton of value so I removed it. I can add something back in though if you'd like.

martinhpedersen commented on mps.go:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	switch cmd, args := shiftArgs(args); cmd {
+	case "list":
```
I think we should keep a local copy of the current list, so the user can review it when they are offline. I imagine there will be a lot of users that don't keep note of their pickup stations.

We could for instance attempt to fetch from the API, write to disk if successful, and fallback to reading from disk. 

What do you think?

martinhpedersen replied:
Thats a good point. It's better to direct users to use INQUIRY when we have that in place 👍

martinhpedersen commented on CONTRIBUTING.md:
```diff
@@ -26,6 +26,7 @@ We welcome your contributions.
 To make the process as seamless as possible, we ask for the following:
 
   - Go ahead and fork the project and make your changes. We encourage pull requests to discuss code changes.
+  - Base your changes off the "develop" branch, not master
```
Thank you! And apologies for not mentioning this sooner 😊

jboverfelt commented on internal/cmsapi/mps.go:
```diff
@@ -0,0 +1,123 @@
+	PathMPSAdd    = "/mps/add"
+	PathMPSDelete = "/mps/delete"
+	PathMPSGet    = "/mps/get"
+	PathMPSList   = "/mps/list"
```
Yes please reach out to them when you get a chance. That endpoint does look better (and has no listed rate limit), but it's hard to say for sure since the Pat key isn't authorized for that endpoint. 

It looks like Express does what we're doing currently but it's impossible to say for sure.

martinhpedersen commented:
@jboverfelt - Haha, yes I noticed. I need to get a U.S license to beat you 😀 

That's perfect. Thank you so much 😊

jboverfelt commented:
Based on Phil and Lee's email I think we should go with the hybrid station endpoint

martinhpedersen commented:
Yes, I agree. 

I have merged my refactor branch now, so if you want to continue working on this then please go ahead 😊 I think we can just continue with the open #51 issue.

Highly appreciate your contribution!

martinhpedersen commented:
You'll find that I have moved MPSHandle and related functions into `cli/mps.go`. The new project structure is not perfect, but I think it's better no structure as it was before ( 🙈 ). I've split the project into three main packages: `cli`, `api` and `app`. The idea is to have core functionality in `app` to be re-used between the cli and api.
